cpmpy/transformations/normalize.py

```diff
@@ -34,3 +37,96 @@ def unravel(lst, append):
     unravel((cpm_expr,), append)
 
     return newlist
+
+
+def simplify_boolean(lst_of_expr, num_context=False):
+    """
+    removes boolean constants from all CPMpy expressions
+    only resulting boolean constant is literal 'false'
+    - list_of_expr: list of CPMpy expressions
+    """
+
+    newlist = []
+    for expr in lst_of_expr:
+
+        if isinstance(expr, bool):
+            # not sure if this should happen here or at construction time
+            expr = BoolVal(expr)
+
+        if isinstance(expr, BoolVal):
+            newlist.append(int(expr.value()) if num_context else expr)
+
+        elif isinstance(expr, Operator):
+            args = simplify_boolean(expr.args, num_context=not expr.is_bool())
+
+            if expr.name == "or":
+                if any(is_true_cst(arg) for arg in args):
+                    newlist.append(1 if num_context else BoolVal(True))
+                else:
+                    newlist.append(Operator("or", [arg for arg in args if not isinstance(arg, BoolVal)]))
+
+            elif expr.name == "and":
+                if any(is_false_cst(arg) for arg in args):
+                    newlist.append(0 if num_context else BoolVal(False))
+                else:
+                    newlist.append(Operator("and", [arg for arg in args if not isinstance(arg, BoolVal)]))
+
+            elif expr.name == "->":
+                cond, bool_expr = args
+                if is_false_cst(cond) or is_true_cst(bool_expr):
+                    newlist.append(BoolVal(True))
+                elif is_true_cst(cond):
+                    newlist.append(bool_expr)
+                elif is_false_cst(bool_expr):
+                    newlist += simplify_boolean([~cond])
+                else:
+                    newlist.append(cond.implies(bool_expr))
+
+            elif expr.name == "not":
+                if is_true_cst(args[0]):
+                    newlist.append(BoolVal(False))
+                elif is_false_cst(args[0]):
+                    newlist.append(BoolVal(True))
+                else:
+                    newlist.append(~args[0])
+
+            else: # numerical expressions
+                newlist.append(Operator(expr.name, args))
+
+        elif isinstance(expr, Comparison):
+            lhs, rhs = simplify_boolean(expr.args, num_context=True)
+            if (isinstance(lhs, int) and isinstance(rhs, _BoolVarImpl)) or isinstance(lhs, BoolVal) and isinstance(rhs, _IntVarImpl):
+                lhs, rhs = rhs, lhs
+            if isinstance(lhs, _BoolVarImpl) and isinstance(rhs, int):
```

Comment:
The central idea is that a comparison of a Boolean variable to an integer value can always be simplified to a simple Boolean expression. E.g., `p <= 0 simplifies to `~p`.